src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -0,0 +1,5130 @@
+    }
+    return os;
+  };
+
+  C_DiscardRequest(T &rwl, const utime_t arrived, Extents &&image_extents,
+		   const int skip_partial_discard, Context *user_req)
+    : C_BlockIORequest<T>(rwl, arrived, std::move(image_extents), bufferlist(), 0, user_req),
+    m_skip_partial_discard(skip_partial_discard) {
+    if (RWL_VERBOSE_LOGGING) {
+      ldout(rwl.m_image_ctx.cct, 99) << this << dendl;
+    }
+  }
+
+  ~C_DiscardRequest() {
+    if (RWL_VERBOSE_LOGGING) {
+      ldout(rwl.m_image_ctx.cct, 99) << this << dendl;
+    }
+  }
+
+  template <typename... U>
+  static inline std::shared_ptr<C_DiscardRequest<T>> create(U&&... arg)
+  {
+    return SharedPtrContext::create<C_DiscardRequest<T>>(std::forward<U>(arg)...);
+  }
+
+  auto shared_from_this() {
+    return shared_from(this);
+  }
+
+  void finish_req(int r) {}
+
+  bool alloc_resources() override;
+
+  void deferred_handler() override { }
+
+  void dispatch() override;
+
+  const char *get_name() const override {
+    return "C_DiscardRequest";
+  }
+};
+
+/**
+ * This is the custodian of the BlockGuard cell for this compare and write. The
+ * block guard is acquired before the read begins to guarantee atomicity of this
+ * operation.  If this results in a write, the block guard will be released
+ * when the write completes to all replicas.
+ */
+template <typename T>
+struct C_CompAndWriteRequest : public C_WriteRequest<T> {
+  using C_BlockIORequest<T>::rwl;
+  bool m_compare_succeeded = false;
+  uint64_t *m_mismatch_offset;
+  bufferlist m_cmp_bl;
+  bufferlist m_read_bl;
+  friend std::ostream &operator<<(std::ostream &os,
+				  const C_CompAndWriteRequest<T> &req) {
+    os << (C_WriteRequest<T>&)req
+       << "m_cmp_bl=" << req.m_cmp_bl << ", "
+       << "m_read_bl=" << req.m_read_bl << ", "
+       << "m_compare_succeeded=" << req.m_compare_succeeded << ", "
+       << "m_mismatch_offset=" << req.m_mismatch_offset;
+    return os;
+  };
+
+  C_CompAndWriteRequest(T &rwl, const utime_t arrived, Extents &&image_extents,
+			bufferlist&& cmp_bl, bufferlist&& bl, uint64_t *mismatch_offset,
+			int fadvise_flags, Context *user_req)
+    : C_WriteRequest<T>(rwl, arrived, std::move(image_extents), std::move(bl), fadvise_flags, user_req),
+    m_mismatch_offset(mismatch_offset), m_cmp_bl(std::move(cmp_bl)) {
+    if (RWL_VERBOSE_LOGGING) {
+      ldout(rwl.m_image_ctx.cct, 99) << this << dendl;
+    }
+  }
+
+  ~C_CompAndWriteRequest() {
+    if (RWL_VERBOSE_LOGGING) {
+      ldout(rwl.m_image_ctx.cct, 99) << this << dendl;
+    }
+  }
+
+  template <typename... U>
+  static inline std::shared_ptr<C_CompAndWriteRequest<T>> create(U&&... arg)
+  {
+    return SharedPtrContext::create<C_CompAndWriteRequest<T>>(std::forward<U>(arg)...);
+  }
+
+  auto shared_from_this() {
+    return shared_from(this);
+  }
+
+  void finish_req(int r) override {
+    if (m_compare_succeeded) {
+      C_WriteRequest<T>::finish_req(r);
+    } else {
+      utime_t now = ceph_clock_now();
+
```

Comment:
I don't think this check (m_perfcount is null not or ) is needed here. It is initialized and no check in other places.

Reply:
I added the comments at line 4293 in ReplicatedWriteLog<I>::flush() , not sure why the display here is line 1788.

Reply:
> I added the comments at line 4293 in ReplicatedWriteLog::flush() , not sure why the display here is line 1788.

It’s shown with the wrong line here, but when you open the diffs for the file it’s on the right line.  It may be unhappy about files this large, or about the updates I’ve pushed since this PR was created.

Reply:
> I don't think this check (m_perfcount is null not or ) is needed here. It is initialized and no check in other places.

The idea was for the perf counters to be optional. If there’s no way to skip creating the perf counter object, there should be. It may only be a build option, but it should be a runtime option. I’m not worried about the performance effect of reading a pointer here. The object memory allocation overhead is much larger than that, and more variable. If we were worried about this, we could make the perfcounter a template argument (const) so the compiler would generate a version of the function with the perf counter clause, and one without.

Reply:
I agree that if it is optional, it should be a runtime option. It means customers can change the option when the service is running. I proposed the above comments as I think we should use logger consistently in RWL. Currently, perf is always enabled, so no need check. 
To make perfcounter a template argument, can it be changed in runtime?
For reference, there is perf and perf_rocksdb options in common/options.cc, but I don't find the usage of perf. Meanwhile, perf_rocksdb is only used for rocksdb inside stats CreateDBStatistics, not used for controlling perfcounter.

Reply:
I remember now. This is because the ImageCache::flush() method might be called by librbd before the cache initializes, or when initialization fails. If you take that out, you should see a segfault near there during the librbd unit test.